### PR TITLE
Add IBM Power VS: machine assets

### PR DESCRIPTION
**Pull request by clnperez**
For more background on IPI on Power VS, refer to the enhancement
proposal here: openshift/enhancements#736

Older discussions on some of the code here can be found in #5224

Signed-off-by: Christy Norman <christy@linux.vnet.ibm.com>

### CODE REVIEW

File: pkg/asset/machines/powervs/machines.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	if platform.ClusterOSImage != "" {
+		image = platform.ClusterOSImage
+	} else {
+		image = fmt.Sprintf("rhcos-%s", clusterID)
+	}
+	if platform.PVSNetworkName != "" {
+		network = platform.PVSNetworkName
+	} else {
+		network = fmt.Sprintf("pvs-net-%s", clusterID)
+	}
```
[the remark this answers is not in the document]
Comment (clnperez) [reply pkg/asset/machines/powervs/machines.go:42]: thanks. some of this has been through many iterations. great to have fresh eyes on it

File: pkg/asset/machines/powervs/machinesets.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	if platform.ClusterOSImage != "" {
+		image = platform.ClusterOSImage
+	} else {
+		image = fmt.Sprintf("rhcos-%s", clusterID)
+	}
+	if platform.PVSNetworkName != "" {
+		network = platform.PVSNetworkName
+	} else {
+		network = fmt.Sprintf("pvs-net-%s", clusterID)
+	}
```
Comment (rna-afk) [inline pkg/asset/machines/powervs/machinesets.go:42]: Need this change here too
```suggestion
	image = fmt.Sprintf("rhcos-%s", clusterID)
	network = fmt.Sprintf("pvs-net-%s", clusterID)
	if platform.ClusterOSImage != "" {
		image = platform.ClusterOSImage
	}
	if platform.PVSNetworkName != "" {
		network = platform.PVSNetworkName
	}
```

**Comment by clnperez** [comment]
@rna-afk does anyone else need eyes on the rest of these?

**Comment by rna-afk** [comment]
I cannot approve these PRs so I'll make sure to review it before I ask one of the team members to approve it.

File: pkg/asset/machines/powervs/machines.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	mpool := pool.Platform.PowerVS
+
+	var (
+		image, network string
+	)
+
+	// Only the service instance is guaranteed to exist and be passed via the install config
+	// The other two, we should standardize a name including the cluster id.
+	image = fmt.Sprintf("rhcos-%s", clusterID)
+	network = fmt.Sprintf("pvs-net-%s", clusterID)
```
Comment (rna-afk) [inline pkg/asset/machines/powervs/machines.go:34]:
```suggestion
	// Only the service instance is guaranteed to exist and be passed via the install config
	// The other two, we should standardize a name including the cluster id.
	image := fmt.Sprintf("rhcos-%s", clusterID)
	network := fmt.Sprintf("pvs-net-%s", clusterID)
```

File: pkg/asset/machines/powervs/machines.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	"github.com/openshift/installer/pkg/types"
+	"github.com/openshift/installer/pkg/types/powervs"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// Machines returns a list of machines for a machinepool.
+func Machines(clusterID string, config *types.InstallConfig, pool *types.MachinePool, role, userDataSecret string) ([]machineapi.Machine, error) {
+
```
Comment (rna-afk) [inline pkg/asset/machines/powervs/machines.go:20]: Do we need this here?
```suggestion
if configPlatform := config.Platform.Name(); configPlatform != powervs.Name {
	return nil, fmt.Errorf("non-powerVS configuration: %q", configPlatform)
}
```

File: pkg/asset/machines/worker.go [hunk 5]
```diff
@@ -158,6 +162,15 @@ func defaultVSphereMachinePoolPlatform() vspheretypes.MachinePool {
 	}
 }
 
+func defaultPowerVSMachinePoolPlatform() powervstypes.MachinePool {
+	return powervstypes.MachinePool{
+		Memory:     "32",
+		Processors: "0.5",
+		ProcType:   "shared",
```
Comment (rna-afk) [inline pkg/asset/machines/worker.go:169]: Probably want to update this to the enum here?

Comment (clnperez) [reply pkg/asset/machines/worker.go]: ah, yah thanks

**Comment by clnperez** [comment]
@rna-afk made all suggested changes

**Comment by rna-afk** [comment]
/lgtm
cc @staebler for approval

**Comment by kirankt** [comment]
Please re-verify vendoring.

File: pkg/asset/machines/master.go [hunk 6]
```diff
@@ -417,6 +421,19 @@ func (m *Master) Generate(dependencies asset.Parents) error {
 			return errors.Wrap(err, "failed to create master machine objects")
 		}
 		vsphere.ConfigMasters(machines, clusterID.InfraID)
+	case powervstypes.Name:
+		mpool := defaultPowerVSMachinePoolPlatform()
+		mpool.Set(ic.Platform.PowerVS.DefaultMachinePlatform)
+		mpool.Set(pool.Platform.PowerVS)
+		// Only the service instance is guaranteed to exist and be passed via the install config
+		// The other two, we should standardize a name including the cluster id. At this point, all
+		// we have are names.
```
Comment (staebler) [inline pkg/asset/machines/master.go:430]: Is this the right place for this comment? A similar comment is in `Machines` function. I don't see a context for service instance or "the other two" here.

**Review by staebler — COMMENTED** [review]
Looks fine to me. I'll add my approve. I am removing the lgtm for my comment. @rna-afk, feel free to re-add the lgtm when the comment is addressed or an explained.
/approve
/lgtm cancel

File: pkg/asset/machines/master.go (hunk 6 above)
Comment (clnperez) [reply pkg/asset/machines/master.go:430]: ah no, it isn't. the _other two_ are `image` and `network` in `Machines()` and `Machinesets()`. ill delete this

**Comment by clnperez** [comment]
@patrickdillon assuming these tests all pass after #5612 merges, can either you or @staebler add your `approve`? it looks like Matthew's didn't take from his last comment.

**Comment by clnperez** [comment]
There is no `approve` label though. Do we need one?

**Review by staebler — COMMENTED** [review]
/approve

**Comment by staebler** [comment]
verify-vendor is failing

/lgtm cancel

**Comment by staebler** [comment]
> verify-vendor is failing
> 
> /lgtm cancel

github.com/openshift/installer/cmd/openshift-install imports
	github.com/openshift/library-go/pkg/config/clusteroperator/v1helpers loaded from github.com/openshift/library-go@v0.0.0-20210811133500-5e31383de2a7,
	but go 1.16 would select v0.0.0-20211214183058-58531ccbde67
github.com/openshift/installer/cmd/openshift-install imports
	github.com/openshift/library-go/pkg/route/routeapihelpers loaded from github.com/openshift/library-go@v0.0.0-20210811133500-5e31383de2a7,
	but go 1.16 would select v0.0.0-20211214183058-58531ccbde67

**Pull request merged**